vernac/record.ml

```diff
@@ -325,6 +326,22 @@ let instantiate_possibly_recursive_type ind u ntypes paramdecls fields =
 
 (* We build projections *)
 
+(** Declare projection [ref] over [from] a coercion
+   or a typeclass instance according to [flags]. *)
+(* remove the last argument (it will become alway true) after deprecation phase
+   (started in 8.17, c.f. https://github.com/coq/coq/pull/16230) *)
+let declare_proj_coercion_instance flags ref from poly with_coercion =
+  if with_coercion && flags.Data.pf_coercion then begin
+    let cl = ComCoercion.class_of_global from in
+    ComCoercion.try_add_new_coercion_with_source ref ~local:false ~poly ~nonuniform:false ~reversible:flags.Data.pf_reversible ~source:cl
+  end;
+  if flags.Data.pf_instance then begin
+    let env = Global.env () in
+    let sigma = Evd.from_env env in
+    let info = Typeclasses.{ hint_priority = flags.Data.pf_priority; hint_pattern = None } in
+    Classes.declare_instance ~warn:true env sigma (Some info) SuperGlobal ref
```

Comment:
As remarked by @JasonGross this is a case of superglobal instances that cannot be deactivated via attributes. Maybe it would be the right time to switch the behaviour to export because it's a change of syntax.

Reply:
We can also introduce per-field attributes for record syntax which might be useful for coercion attributes later on. Doing so would allow us to switch to #[export] and keep #[global] as a compat solution.

Reply:
I did the following (c.f. commit [Add locality attribute to record fields for :> and ::](https://github.com/coq/coq/pull/16230/commits/cb0bad032cecb952d14bb5d2eaad234566484017)):
* keep SuperGlobal the default for now (for both the deprecated `:>` and the new `::` syntax)
* add support for locality attribute on record fields (working for both coercions and typeclass instances)
* add a warning in the absence of an explicit attribute

We could alternatively make the new syntax `Export` rather than `SuperGlobal` by default but this means change of syntax would come with a change of semantics without having never seen any warning which would be rude IMHO.

Reply:
@ppedrot @Alizter another alternative would be to make the new `::` syntax `Export` by default and have the deprecation warning for the old `:>` syntax explicitly mention this (something like "Beware that the default locality for `::` is `#[export]` instead of `#[global]` for `:>`. Add an explicit `#[global]` attribute to the field if you want to keep the current behavior.").

Reply:
@proux01 As you have added support for attributes in the field, we can safely switch `::` to mean export now as @ppedrot suggests. Let's make sure to include the locality change in the deprecation message for `:>`. Something like
```
The hint locality of "::" is #[export] by default, as apposed to #[global] for ":>". This should not cause any problems, but if it does then add the #[global] to the "::" field.
```

Perhaps worded a bit better tho.

Reply:
Done